Make minimum size for secure memory a size_t.

paulidale opened the pull request:
The minimum size argument to CRYPTO_secure_malloc_init() was an int but ought
to be a size_t since it is a size.

From an API perspective, this is a change.  However, the minimum size is
verified as being a positive power of two and it will typically be a small
constant, so impact should be low to none.

- [x] documentation is added or updated
- [x] tests are added or updated

mattcaswell commented:
I wonder if there are any compilers which might complain about this if applications make this call? Does this deserve a CHANGE entry...or is it more like the "const" changes we make which don't usually have one.

paulidale commented:
`test/secmem.c` didn't actually have a **size_t** cast -- one in a comment, one inside `#if 0` because it tests an infinite loop condition.  Nonetheless, I've removed the cast from the second argument of both but not from the first where a wider than **int** value is required.

I've updated the documentation.

@mattcaswell, I've added a note in the _history_ section of the documentation.  CHANGES seems excessive but I'm happy to add an entry.

paulidale commented:
Is a vote required for this?

Technically, it is a breaking API change.  Realistically, it's very safe.  Unless, someone is taking the address of the function and assigning it to a variable.

DDvO commented:
> I wonder if there are any compilers which might complain about this if applications make this call? Does this deserve a CHANGE entry...or is it more like the "const" changes we make which don't usually have one.

One potential compile-time problem could be if a caller passes in a clearly negative value, thought is is very unlikely since there was already the dynamic check for negative input, which should have lead to a run-time error.
I've just tried out with `gcc` on code configured with `--strict-warnings`. Interestingly, it does not complain if a `-1` is passed as actual argument and one changes the parameter type from  `int`  to  `size_t`.

richsalz commented:
I think this is a bad thing to do, for a couple of reasons
- It could break something
- It works fine now
- It's a lot of changes (doc, code, casts) for no real gain other than nerd esthetics.
- It doesn't fully address the issue #10995 where this was used as an example.
- It establishes a bad precedent of "we think it safe let's fix it" which is counter to one of 3.0's stated goals: "most applications should work just be recompiling."  The slippery slope is very very slippery. :)

mattcaswell commented:
> a couple of reasons

It amuses me that you list 5 reasons:
https://xkcd.com/1070/

levitte commented:
> It's a lot of changes (doc, code, casts) for no real gain other than nerd esthetics.

I've heard that consistency is a good thing :wink:

richsalz commented:
>I've heard that consistency is a good thing

Okay, I'll take the bait since you're trying to use my words against me:  how does this PR make things more consistent?

richsalz commented:
> It amuses me that you list 5 reasons:

I was sloppy and didn't go back and fix the first line. :)  The bullets just kept a-comin'.

I had an English class once where we had to sort sentence in order:
```
A couple of the clarinetists...
Most of the clarinetists...
Many of the clarinetists...
A few of the clarinetists...
```
Wish I could find that page; it went on for 50 or so lines!

DDvO commented:
The event that a caller provides a negative value for the `minsize` parameter is very unlikely because it make no sense and typically the value will be a simple constant, so would be obvious.
And even if so, what could really go wrong here if that is subsequently taken as a huge unsigned value?
AFAICS:
* a huge about of heap memory is reserved successfully, which is again very unlikely in itself
* the caller obtains an error result, as before with the hitherto version of the function

levitte commented:
> how does this PR make things more consistent?

Easy.  Consistent use of `size_t` for passing sizes.

richsalz commented:
So "fix" SSL_read and SSL_write?  Those are far more glaring inconsistencies.  And fixing this one, which was done deliberately and correctly... shrug.  I've said why I think this is a bad idea and given explanations.  Oh well.

levitte commented:
Actually, I see no reason why we shouldn't fix `SSL_read` and `SSL_write` in a similar way...
However, that shouldn't be a factor in approving or not approving this PR.

DDvO commented:
As there are concerns on doing this and related changes right now, how about simply deferring them to after 3.0? 
There is nothing urgent in here.

paulidale commented:
Major releases are the time for changes like this -- API compatibility isn't guaranteed at these points.  A deferral until after 3.0 will be a long time after...

Changing the _int_ to a _size_t_ is definitely break ABI compatibility.

davxy commented:
Well... I'm not a committer so my opinion is not so influent. But I think that if you wait after 3.0 for stuff like this, the change will never happen.

A major release like 3.0 is not only an opportunity to massively change the internals, but is also the best opportunity to improve the API wherever is required.

Is not only a matter of nerd esthetics but is a matter of type safety and coerence.

As I understood, for someone (maybe the original author) these are stupid changes, but for the community means a lot to have a consistent api.

DDvO commented:
> Major releases are the time for changes like this -- API compatibility isn't guaranteed at these points. A deferral until after 3.0 will be a long time after...

Now I'm confused. @kroeckx explained to me a couple of months ago that API changes for 3.0 should be kept to the minimum: https://github.com/openssl/openssl/pull/10504#issuecomment-558249028
> OpenSSL 3.0 is supposed to remain as compatible as possible, we expect most applications to just need to rebuild and not make any changes. Anything that requires them to make changes needs to wait until at least 4.0.

paulidale commented:
Major releases are when we _can_ make changes.  For 3.0, the decision was made that we are aiming for a high level of source compatibility and this _will_ limit changes.

I can't see this particular change breaking anyone.

paulidale commented:
Merged to master.  Thanks for the discussion.
